Mathlib/Algebra/DirectSum/Internal.lean

```diff
@@ -428,3 +438,78 @@ theorem SetLike.Homogeneous.smul [CommSemiring S] [Semiring R] [Algebra S R] {A
   ⟨i, Submodule.smul_mem _ _ hi⟩
 
 end HomogeneousElement
+
+/-! ### ℕ-gradings -/
+
+section Semiring
+
+variable [Semiring R] [SetLike σ R] [AddSubmonoidClass σ R]
+variable {A : ℕ → σ} [SetLike.GradedMonoid A]
+
+theorem mul_apply_eq_zero {r r' : ⨁ i, A i} {m n : ℕ}
+    (hr : ∀ i < m, r i = 0) (hr' : ∀ i < n, r' i = 0) ⦃k : ℕ⦄ (hk : k < m + n) :
+    (r * r') k = 0 := by
+  rw [Subtype.ext_iff, ZeroMemClass.coe_zero, coe_mul_apply_eq_sum_antidiagonal]
+  apply Finset.sum_eq_zero fun x hx ↦ ?_
+  obtain (hx | hx) : x.1 < m ∨ x.2 < n := by simp at hx; omega
+  all_goals simp [hr, hr', hx]
```

Comment:
This looks quite similar to some theorems we have about `MonoidAlgebra`, which are stated for more general classes of indices than just `Nat`. Can we write the same lemmas?

Reply:
I did actually try that, but I needed `LinearOrderedAddMonoid`, `Finset.HasAntidiagonal`, and at one point even `CanonicallyOrderedAdd`, and then only type we have satisfying all of these is `ℕ`, so I specialized.

However, with your above comment about `coe_mul_apply`, I'll revisit this.